### DISCUSSION TITLE
Release as a container?

### DISCUSSION THREAD

**Discussion by jerabaul29** [helix-editor/helix#7004] [General] [answered]
Would it be possible to release helix packaged as a container (singularity / podmain container for example)? That would be very convenient to make it easy to run helix without any dependence on the host, what packages are on it, etc.

For example, I started using helix on my Ubuntu 22.04 computer using the AppImage and it works just fine. But trying to do the same on my Ubuntu 20.04, I get some glibc errors (```/lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.34' not found``` shows I am on ```ldd (Ubuntu GLIBC 2.31-0ubuntu9.9) 2.31```). But I do not want to start hacking around / changing my glibc version (currently ```/lib/x86_64-linux-gnu> ldd --version libc.so```), I did this once in the past and ended up in a world of pain.

By releasing a singularity or podman container that runs helix, this kind of issue would disappear for users like me who need to run on some slightly older systems for some of their activity.

**Comment by jwijenbergh** [comment]
Flatpak https://flathub.org/apps/com.helix_editor.Helix covers your concerns. So does the Nix package

**Comment by jerabaul29** [comment]
Thanks for pointing this out. A few comments:

- Let us say I also want to include some additional "non mandatory dependencies", such as a few LSPs. In a singularity / podman container, I could just grab the definition file, and add the packages I want. Not sure if this is as easy with flatpak? This means that I do not really have a good way to have the LSPs I want being "independent" of the host OS, libraries, etc, and bundled with my helix image?
- I also wonder if the process of building locally a singularity or podman image is maybe easier / more common than building flatpak images locally for testing, expending, etc? At least from a short search there is a huge volume of help / info on the topic for singularity / podman container compared with what is available for flatpak (though this may be a bit subjective too).

**Comment by jerabaul29** [comment]
PS: it also looks like the flatpak installation method is not mentioned in the doc? (at least if I am not missing something, I cannot see it at https://docs.helix-editor.com/install.html ). Maybe it would be useful to mention it there in a new section? :)

**Comment by David-Else** [comment]
@blt-r I only just saw the flatpak, great work! Is it possible to get the Flatpak verified https://flathub.org/en-GB/apps/collection/verified/ ? I think now this new feature is out some people will steer clear of apps that are not verified.

@jerabaul29 Yes, Flatpak definitely needs adding, I have done it: https://github.com/helix-editor/helix/pull/6994

**Comment by jwijenbergh** [comment]
> Let us say I also want to include some additional "non mandatory dependencies", such as a few LSPs. In a singularity / podman container, I could just grab the definition file, and add the packages I want. Not sure if this is as easy with flatpak? This means that I do not really have a good way to have the LSPs I want being "independent" of the host OS, libraries, etc, and bundled with my helix image?
> I also wonder if the process of building locally a singularity or podman image is maybe easier / more common than building flatpak images locally for testing, expending, etc? At least from a short search there is a huge volume of help / info on the topic for singularity / podman container compared with what is available for flatpak (though this may be a bit subjective too).

In my mind, running a simple terminal based text editor in a container is pretty overkill. Have you looked at direnv with nix or https://github.com/numtide/devshell? That gives you dependencies *per* project without cluttering your OS. For my development workflow, the language servers that I use are only available in each project directory. So e.g. a Go project X has gopls and go compilers installed, whereas a python project Y, only has pyright and a python interpreter

Apologies if this is not the answer you're looking for. I'm not sure if you can add dependencies to flatpak, you probably can by forking the package?

Anyways, a full podman container is probably possible but you will have to give so many permissions that the whole "container" aspect is not really useful and just adds complexities

**Comment by jerabaul29** [comment]
@jwijenbergh many thanks for the in-depth discussion :) . You are probably right, and I can try for now the flatpak + non containerized LSPs. If I meet issues getting LSPs to run and need to containerize them, then I can come back to this idea :) .

Regarding the flatpak: I agree having it signed would be great, I do not like installing unsigned / unverified stuff, and I think more and more organizations are becoming more security aware and set stricter requirements on this kind of aspects (a bit similar to https://github.com/helix-editor/helix/issues/6946 , each for their respective distribution way).

**Comment by blt-r** [comment]
> Is it possible to get the Flatpak verified

There are two ways to verify the flatpak. 
1. Through the website. We will need to put a specific file at a specific URL at https://helix-editor.com. I think this is the better way, because the verification checkmark would be prettier.
2. Through GitHub. We will need to give to some maintainer of this repo a write access to https://github.com/flathub/com.helix_editor.Helix (I don't know how). And then they can log into flathub.org with their GitHub account, go to developer settings and it will somehow detect that they are maintainer of the repo and allow them to just verify it. I'm not sure how exactly this works, but I can ask on flathub forums.

**Comment by David-Else** [comment]
> There are two ways to verify the flatpak.
> 
>     1. Through the website. We will need to put a specific file at a specific URL at https://helix-editor.com. I think this is the better way, because the verification checkmark would be prettier.
> 

@blt-r Would you be OK to provide the file? @the-mikedavis Does this sound good to you?

**Comment by blt-r** [comment]
We need to place 
```
88a85820-7ec3-4c11-94ce-3e736dfda445
```
at
```
https://helix-editor.com/.well-known/org.flathub.VerifiedApps.txt
```

**Comment by ghost** [comment]
@jerabaul29 the dockerfile for your issue is simple

use archlinux:latest as it has latest helix/glibc in repos

```dockerfile
FROM archlinux:latest
RUN sudo pacman -Sy helix --noconfirm
RUN echo "helix installed"
EXPOSE 8080
```

This Dockerfile builds an image of archlinux with helix and run these below commands:-

```
docker build -t helix_container
docker run -d -it -p 80:8080 --name=helix_container helix -- --host=0.0.0.0 
```

**Comment by jerabaul29** [comment]
Thanks @sontaimnt ; question: why do you want to expose 8080?

**Reply by ghost** [reply]
Thats what i generally expose... i am new to Docker so this is what they did for the tutorial

**Comment by blt-r** [comment]
There were people in this thread asking about adding language serves to the flatpak container. Flatpak has SDK extensions for doing that. And the Helix flatpak uses [ide-flatpak-wrapper](https://github.com/flathub/ide-flatpak-wrapper) to make it easier to automatically enable the SDK extensions. I wrote instructions for doing that in [helix_first_run.txt](https://github.com/flathub/com.helix_editor.Helix/blob/master/helix_first_run.txt). That file should also automatically open the first time you run the flatpak.

**Comment by the-mikedavis** [comment]
We don't maintain the flatpak "officially" (i.e. none of the maintainers are involved with the package) so verification doesn't make sense in this case.

We try to keep the set of official packages small to limit our maintenance burden when cutting a release. It's easy to create a container based on the binary builds attached to github releases, those in package managers or "from scratch" (building it with Rust and C/C++ toolchains) so I don't think we need to produce an official container image.

**Comment by jerabaul29** [comment]
Coming back to this: I am trying to "build from scratch" in order to be able to develop a bit, but I am hitting issues (in particular, some issues similar to https://github.com/rust-lang/cc-rs/issues/909 ). Wondering if it could be useful to provide a dockerfile that is able to build the project from source, so that this can be used as a common development environment where "everything should work"?

**Reply by kirawi** [reply]
IMO you should try using nix. It addresses that issue with reproducible builds and it has first-class support from us.

**Reply by jerabaul29** [reply]
So that would mean installing the helix package from https://nixos.org/ , right? In case I want to develop / edit the code and compile it locally, would this be something nix would also help with?

**Reply by kirawi** [reply, marked as answer]
Yes, https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-develop

You only need to install the package manager itself.

**Comment by tfmoraes** [comment]
Other option is to use [Linuxbrew ](https://docs.brew.sh/Homebrew-on-Linux).
